### DISCUSSION TITLE
Can't replicate kdm0

### DISCUSSION THREAD

**Discussion by drstrupf** [dayoonkwon/BioAge#40] [Q&A] [answered]
Dear Dayoon Kwon,

thank you for this nice package! I've successfully used the PhenoAge functionality to create custom models, and am now looking into KDM.  As a first step, I repeated the training on NHANES 3 you did when preparing the data sets (lines 944 and following in [nhanes_all.R](https://github.com/dayoonkwon/BioAge/blob/master/data-raw/nhanes_all.R)). I.e. I used the same set of biomarkers and filter conditions for training:

`train_fem_kdm0 = kdm_calc(data = NHANES3 %>% filter(gender==2 & age>=30 & age<=75 & pregnant==0), c("fev","sbp","totchol","hba1c","albumin","creat","lncrp","alp","bun"), fit = NULL, s_ba2 = NULL)`

However, in `train_fem_kdm0$data` I then found that `kdm` and `kdm0` are not equal, with `kdm = (kdm0 - age) * x + age`, where x is approximately 2.175 (min. 2.171, mean 2.175, max. 2.180). So this factor only *almost* constant between subjects. 

Could you please help me understand what I'm doing wrong?

Thank you very much, and best regards
Ste

**Comment by dayoonkwon** [comment]
`kdm0` and `kdm` are expected to differ. `kdm0` was pre-computed using the original Levine 2013 algorithm with fixed published coefficients. `kdm_calc` re-estimates all parameters (k, q, s, s_ba2) from your training data. Even with the same biomarker set, the re-estimated parameters won't be identical to the original published values, which produces the scaling difference you're seeing. This is by design — the package allows you to train new versions of the algorithm rather than only reproducing the original.

**Comment by drstrupf** [comment]
Dear Dayoon Kwon,

Thank you for your reply, and sorry, I did not phrase my question properly. I was specifically asking about ``kdm0`` in the NHANES data you provided. In these data files, you provide the columns ``phenoage0`` and ``kdm0``, which are described as computed using the original algorithms, i.e. Levine 2013 for KDM. 

Now for PhenoAge, these published coefficients are hard-coded in [nhanes_all.R](https://github.com/dayoonkwon/BioAge/blob/b1f9fc02f086cd4aa74185f2335ab1366082e7fe/data-raw/nhanes_all.R#L972) from line 972 onwards and I can replicate the column ``phenoage0``. 

However, the calculation of ``kdm0`` in [nhanes_all.R](https://github.com/dayoonkwon/BioAge/blob/b1f9fc02f086cd4aa74185f2335ab1366082e7fe/data-raw/nhanes_all.R#L943) from line 943 onwards is a re-training (no hard-coded coefficients), and if I follow these exact steps, I can *not* reproduce the ``kdm0`` column in the provided NHANES datasets, even though it is supposed to be calculated by this block of code.

Thus I was wondering what the column ``kdm0`` in the NHANES data files is, and I would be grateful if you could help me understand the dfference between the ``kdm0`` in the data files and the ``kdm0`` computed in [nhanes_all.R](https://github.com/dayoonkwon/BioAge/blob/b1f9fc02f086cd4aa74185f2335ab1366082e7fe/data-raw/nhanes_all.R#L943) from line 943 onwards.

Thank you & kind regards
Ste

**Reply by dayoonkwon** [reply, marked as answer]
The `kdm0` values in the provided NHANES datasets were computed with an earlier version of `kdm_calc` that handled missing biomarkers differently. The current version of `kdm_calc` includes a rescaling step for missing biomarkers that was not present when `kdm0` was originally computed. This is why re-running `nhanes_all.R` with the current code produces slightly different values. The underlying algorithm and biomarker set are the same — the difference is in how participants with incomplete data are handled during the variance (`s_ba2`) estimation.

**Reply by drstrupf** [reply]
Ah I see, thank you!
